src/main/kotlin/uk/gov/justice/digital/hmpps/approvedpremisesapi/service/cas1/planning/SpaceBookingDayPlanner.kt

```diff
@@ -0,0 +1,129 @@
+   *  If a room has multiple beds, the booking will be assigned to all beds</li>
+   * </ol>
+   *
+   * @param beds All beds available on the day being planned (i.e. won't include out of service beds or beds inactive according to start/end date)
+   * @param bookings All required bookings. The code assumes that the premise level characteristics have already been met
+   */
+  fun plan(
+    beds: Set<Bed>,
+    bookings: Set<SpaceBooking>,
+  ): DayPlannerResult {
+    val bedLedger = BedLedger(beds)
+    val planned = mutableListOf<DayPlannerPlannedBooking>()
+    val unplanned = mutableListOf<DayPlannerUnplannedBooking>()
+
+    val sortedBookings = bookings.toList()
+      .sortedByDescending { b -> b.requiredCharacteristics.sumOf { c -> c.weighting } }
+      .toMutableList()
+
+    sortedBookings.forEach { booking ->
+      val requiresSingleRoom = booking.requiredCharacteristics.any { it.singleRoom }
+      val characteristicsExcludingSingleRoom = booking.requiredCharacteristics.filter { !it.singleRoom }.toSet()
+
+      when (
+        val findResult = bedLedger.findBed(
+          characteristics = characteristicsExcludingSingleRoom,
+          requiresSingleRoom = requiresSingleRoom,
+        )
+      ) {
+        is FindBedResult.BedsFound -> {
+          findResult.beds.forEach { bed ->
+            planned.add(
+              DayPlannerPlannedBooking(
+                bed = bed,
+                booking = booking,
+              ),
+            )
+            bedLedger.reserve(bed)
+          }
+        }
+
+        is FindBedResult.BedNotFound -> {
+          unplanned.add(
+            DayPlannerUnplannedBooking(
+              booking = booking,
+            ),
+          )
+        }
+      }
+    }
+
+    return DayPlannerResult(
+      beds = beds,
+      planned = planned.toList(),
+      unplanned = unplanned.toSet(),
+    )
+  }
+}
+
+private class BedLedger(initialState: Set<Bed>) {
+  val availableBeds = initialState.sortedBy { it.label }.toMutableList()
+  val reservedBeds = mutableListOf<Bed>()
+
+  fun findBed(
+    characteristics: Set<Characteristic>,
+    requiresSingleRoom: Boolean,
+  ): FindBedResult {
+    val bedsWithMatchingCharacteristics = availableBeds
+      .filter { availableBed -> availableBed.room.characteristics.containsAll(characteristics) }
+      .filter { availableBed -> !requiresSingleRoom || !roomOccupied(availableBed) }
+
+    val bedWithLeastUnusedCharacteristics = bedsWithMatchingCharacteristics
+      .map { bed -> Pair(bed, bed.room.characteristics.minus(characteristics).size) }
+      .minByOrNull { it.second }?.first
+
+    return if (bedWithLeastUnusedCharacteristics != null) {
+      if (requiresSingleRoom) {
+        FindBedResult.BedsFound(allAvailableBedsInRoom(bedWithLeastUnusedCharacteristics.room))
+      } else {
+        FindBedResult.BedsFound(setOf(bedWithLeastUnusedCharacteristics))
+      }
+    } else {
+      FindBedResult.BedNotFound
+    }
+  }
+
```

Comment:
I may be wrong but as far as I can see, there's nothing to force a person into the fullest multiple occupancy room possible if the booking don't have 'single' set, or, more importantly, to put a single occupancy booking into the room with the smallest number of beds.
This `findBed` routine will be called a lot and has several all-bed loops within it so we're probably talking of a performance going with n^2. If performance is an issue, I think the number of scans in this area could be reduced.

Reply:
Some great points.

> there's nothing to force a person into the fullest multiple occupancy room possible if the booking don't have 'single' set

Because we plan bookings with characteristics first (e.g. bookings requiring single rooms), i don't think we need to explicit add this logic, although I agree that this is definitely a goal of the planning. I'm going to look at documenting the goals of the planning at the top of this class and including this so we don't lose sight of it

I'll also add a unit test to make this expectation clear

> put a single occupancy booking into the room with the smallest number of beds.

This is a really good point - I'll explicitly add this logic and add a unit test

> This findBed routine will be called a lot and has several all-bed loops within it so we're probably talking of a performance going with n^2. If performance is an issue, I think the number of scans in this area could be reduced.

I'm cautiously optimistic that we won't need to optimise given that premises won't have too many beds e.g. <100. We have a series of tickets on the occupancy epic APS-1256 with an end goal of trying to plan  a period of time for a given premise against existing bookings and see what the outcome is. We can capture timings from this. I've updated the ticket to make sure we do that.

Reply:
Done! I've rewritten the documentation on the planning class to reflect the discussion above, and highlight some things to consider/optimise if we ever want to start presenting the 'booking to room' planning to premises.